pkg/activator/throttler_test.go

```diff
@@ -0,0 +1,249 @@
+	}{
+		{
+			label:           "all good",
+			revisionGetter:  existingRevisionGetter(10),
+			endpointsGetter: existingEndpointsGetter,
+			maxConcurrency:  defaultMaxConcurrency,
+			want:            int32(10),
+		},
+		{
+			label:           "unlimited concurrency",
+			revisionGetter:  existingRevisionGetter(0),
+			endpointsGetter: existingEndpointsGetter,
+			maxConcurrency:  100,
+			want:            int32(100),
+		},
+		{
+			label:           "non-existing revision",
+			revisionGetter:  nonExistingRevisionGetter,
+			endpointsGetter: existingEndpointsGetter,
+			maxConcurrency:  defaultMaxConcurrency,
+			want:            int32(0),
+			wantError:       fmt.Sprintf("no revision was found for: %s", revID),
+		},
+	}
+	for _, s := range samples {
+		t.Run(s.label, func(t *testing.T) {
+			var received string
+			want := s.want
+			throttler := getThrottler(s.maxConcurrency, s.revisionGetter, s.endpointsGetter, t)
+			err := throttler.UpdateCapacity(revID, 1)
+			if s.wantError != "" {
+				received = err.Error()
+			}
+			if received != s.wantError {
+				t.Errorf("Expected error in Update capacity. Want %s, got %s", s.wantError, err.Error())
+			}
+			if want > 0 {
+				breaker, _ := throttler.breakers[revID]
+				got := breaker.Capacity()
+				if got != want {
+					t.Errorf("Unexpected capacity of the breaker. Want %d, got %d", want, got)
+				}
+			}
+		})
+	}
+}
+
+func TestThrottler_Try(t *testing.T) {
+	samples := []struct {
+		label           string
+		addCapacity     bool
+		wantBreakers    int32
+		wantError       string
+		revisionGetter  func(RevisionID) (*v1alpha12.Revision, bool, error)
+		endpointsGetter func(RevisionID) (int32, error)
+	}{
+		{
+			label:           "all good",
+			addCapacity:     true,
+			wantBreakers:    int32(1),
+			wantError:       "",
+			revisionGetter:  existingRevisionGetter(10),
+			endpointsGetter: existingEndpointsGetter,
+		},
+		{
+			label:           "non-existing revision",
+			addCapacity:     true,
+			wantBreakers:    int32(0),
+			wantError:       fmt.Sprintf("no revision was found for: %s", revID),
+			revisionGetter:  nonExistingRevisionGetter,
+			endpointsGetter: existingEndpointsGetter,
+		},
+		{
+			label:           "non-existing endpoint",
+			addCapacity:     false,
+			wantBreakers:    int32(0),
+			wantError:       sampleError,
+			revisionGetter:  existingRevisionGetter(10),
+			endpointsGetter: nonExistingEndpointsGetter,
+		},
+	}
+	for _, s := range samples {
+		t.Run(s.label, func(t *testing.T) {
+			var got int32
+			var received string
+			want := s.wantBreakers
+			throttler := getThrottler(defaultMaxConcurrency, s.revisionGetter, s.endpointsGetter, t)
+			if s.addCapacity {
+				throttler.UpdateCapacity(revID, 1)
+			}
+			err := throttler.Try(revID, func() {
+				got++
+			})
+			if s.wantError != "" {
+				received = err.Error()
+			}
+			if received != s.wantError {
+				t.Errorf("Expected error in the Try. Want %s, got %s", s.wantError, received)
+			}
```

Comment:
Can those cases be folded? Like:

```golang
if s.wantError != ""; gotError := err.Error() && got != s.wantError {
  t.Errorf("Expected error in the Try. Want %s, got %s", s.wantError, gotError)
}
```

Same for the test above. You can then drop the extra `received` variable and dense up this block quite a bit.

Reply:
Personally, I'd always reduce the scope of variables to be as narrow as possible and try to think of them as constants as much as possible (like immutables in Scala for example).

In the current construct my head needs to parse:

1. received is defined somewhere in the outer scope (close to the top, far away from its usage)
2. It is set in an inner scope (if) but not used there
3. It is used in the outer scope (if) again

Basically I need to track and trace those in my head spanning 13 LoC vs. having them all being defined in 2 LoC.

However, I don't feel particularly strongly about it in tests. If you'd like to keep it that's fine with me.

Reply:
I wouldn't say that reading a single line of code could be a less conundrum then jumping between them. I'm fine with applying the latest suggestion.